### PR TITLE
Implement Shield (wip)

### PR DESCRIPTION
**Pull request by pandaaaBE**
## Introduction
This PR adds working shield.

this will have a lot of commits because i am working on 2 different desktops and this is the best way for me to save the files sry bros

### API changes
`VanillaItems::SHIELD()`
`Item::canDisableShield()` in Item.php & Axe.php

## Tests

I tested this PR by doing the following (tick all that apply):
- [ ] Playtesting using a Minecraft client (provide screenshots or a video)

### CODE REVIEW

File: src/entity/Human.php [hunk 1]
```diff
@@ -546,4 +550,35 @@ protected function destroyCycles() : void{
 		);
 		parent::destroyCycles();
 	}
+
+	protected function onBlock(Entity $entity, EntityDamageEvent $e) : void{
+		parent::onBlock($entity, $e);
+
+		$shield = $this->getInventory()->getItemInHand();
+		$shieldOffhand = $this->getOffhandInventory()->getItem(0);
```
Comment (JavierLeon9966) [inline src/entity/Human.php]:
```suggestion
		$shieldOffhand = $this->getOffHandInventory()->getItem(0);
```

File: src/entity/Living.php [hunk 2]
```diff
@@ -685,10 +684,14 @@ protected function entityBaseTick(int $tickDiff = 1) : bool{
 					}
 				}
 			}
+
```
Comment (JavierLeon9966) [inline src/entity/Living.php:705]:
```suggestion
```

File: src/entity/Living.php [hunk 3]
```diff
@@ -920,4 +924,70 @@ protected function destroyCycles() : void{
...
+		if($event->isCancelled()){
+			return false;
+		}
+
+		if($damager instanceof Living){
+			$deltaX = $damager->getPosition()->x - $this->getPosition()->x;
+			$deltaZ = $damager->getPosition()->z - $this->getPosition()->z;
+			$damager->knockBack($deltaX, $deltaZ);
+			$damager->attackTime = 10;
+			$damager->isAttackTimeByShieldKb = true;
```
Comment (JavierLeon9966) [inline src/entity/Living.php]: Why do we need to reset the damager's invincibility ticks and flag if it was by shield? This doesn't seem necessary

File: src/entity/Living.php [hunk 4]
```diff
@@ -920,4 +924,70 @@ protected function destroyCycles() : void{
...
+	public function isBlocking() : bool{
+		return $this->isBlocking;
+	}
+
+	public function setBlocking(bool $value) : void{
+		$this->getNetworkProperties()->setGenericFlag(EntityMetadataFlags::BLOCKING, $value);
+		$this->isBlocking = $value;
+	}
+
+	public function preAttack(Player $player) : void{
```
Comment (JavierLeon9966) [inline src/entity/Living.php]: This is unused. What is this for?

File: src/entity/Living.php [hunk 5]
```diff
@@ -920,4 +924,70 @@ protected function destroyCycles() : void{
...
+		if($this->isAttackTimeByShieldKb){
+			$this->attackTimeBefore = $this->attackTime;
+			$this->attackTime = 0;
+		}
+	}
+
+	public function postAttack(Player $player) : void{
```
Comment (JavierLeon9966) [inline src/entity/Living.php]: Same for this.

File: src/event/entity/EntityDamageBlockedEvent.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+declare(strict_types=1);
+
+namespace pocketmine\event\entity;
+
+use pocketmine\entity\Entity;
+use pocketmine\event\CancellableTrait;
+/**
+ * @phpstan-extends EntityDamageByEntityEvent<Entity>
+ */
+class EntityDamageBlockedEvent extends EntityDamageByEntityEvent{
```
Comment (JavierLeon9966) [inline src/event/entity/EntityDamageBlockedEvent.php]: This class seems useless when you can just do:
```php
$ev = new EntityDamageByEntityEvent($damager, $entity, EntityDamageEvent::CAUSE_BLOCKING, $damage);
```

File: src/item/Shield.php [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+	public function getMaxDurability() : int{
+		return 336;
+	}
+
+	public function getMaxStackSize() : int{
+		return 1;
+	}
+
+	public function getCooldownTicks() : int{
```
Comment (JavierLeon9966) [inline src/item/Shield.php]: Shields have no interaction cooldown.

File: src/player/Player.php [hunk 8]
```diff
@@ -2710,4 +2711,25 @@ public function onChunkUnloaded(int $chunkX, int $chunkZ, Chunk $chunk) : void{
...
+			$this->setBlocking($shouldBlock);
+		}
+	}
+
+	protected function onBlock(Entity $entity, EntityDamageEvent $e) : void{
+		parent::onBlock($entity, $e);
+
+		if($e->isBreakShield()){
+			$this->resetItemCooldown($this->getInventory()->getItemInHand());
+			$this->resetItemCooldown($this->getOffHandInventory()->getItem(0));
```
Comment (JavierLeon9966) [inline src/player/Player.php]: The cooldown is supposed to be for blocking, not interacting.

**Review by JavierLeon9966 — CHANGES_REQUESTED** [review]
Instead of naming the methods and properties `breakShield` it should be like "shield disabled". `canDisableShield`, `isShieldDisabled`, `getShieldDisableCooldown`, etc.

**Comment by pandaaaBE** [comment]
Closed because the current code is probably not good enough

**Pull request closed without merging**